src/timely-util/src/builder_async.rs

```diff
@@ -212,6 +219,110 @@ impl<'handle, T: Timestamp, D: Container, P: Pull<BundleCore<T, D>>> Future
     }
 }
 
+// TODO: delete and use CapabilityTrait instead once TimelyDataflow/timely-dataflow#512 gets merged
+pub trait CapabilityTrait<T: Timestamp> {
+    fn session<'a, D, P>(
+        &'a self,
+        handle: &'a mut OutputHandleCore<'_, T, D, P>,
+    ) -> Session<'a, T, D, PushCounter<T, D, P>>
+    where
+        D: Container,
+        P: Push<BundleCore<T, D>>;
+}
+
+impl<T: Timestamp> CapabilityTrait<T> for InputCapability<T> {
+    fn session<'a, D, P>(
+        &'a self,
+        handle: &'a mut OutputHandleCore<'_, T, D, P>,
+    ) -> Session<'a, T, D, PushCounter<T, D, P>>
+    where
+        D: Container,
+        P: Push<BundleCore<T, D>>,
+    {
+        handle.session(self)
+    }
+}
+
+impl<T: Timestamp> CapabilityTrait<T> for Capability<T> {
+    fn session<'a, D, P>(
+        &'a self,
+        handle: &'a mut OutputHandleCore<'_, T, D, P>,
+    ) -> Session<'a, T, D, PushCounter<T, D, P>>
+    where
+        D: Container,
+        P: Push<BundleCore<T, D>>,
+    {
+        handle.session(self)
+    }
+}
+
+pub struct AsyncOutputHandle<T: Timestamp, D: Container, P: Push<BundleCore<T, D>> + 'static> {
+    handle: Rc<RefCell<OutputHandleCore<'static, T, D, P>>>,
+    wrapper: Rc<Pin<Box<OutputWrapper<T, D, P>>>>,
+}
+
+impl<T, D, P> AsyncOutputHandle<T, D, P>
+where
+    T: Timestamp,
+    D: Container,
+    P: Push<BundleCore<T, D>> + 'static,
+{
+    fn new(wrapper: OutputWrapper<T, D, P>) -> Self {
+        let mut wrapper = Box::pin(wrapper);
+        // SAFETY:
+        // get_unchecked_mut is safe because we are not moving the wrapper
+        //
+        // transmute is safe because:
+        // * We're erasing the lifetime but we guarantee through field order that the handle will
+        //   be dropped before the wrapper, thus manually enforcing the lifetime.
```

Comment:
Maybe add a comment on `AsyncOutputHandle` that the field order must not be changed, or add a `Drop` impl?